pwndbg/gdblib/functions.py

```diff
@@ -136,6 +136,134 @@ def base(name_pattern: gdb.Value | str):
+    0x555555559010:	0x00	0x00	0x00	0x00	0x00	0x00	0x00	0x00
+    0x555555559018:	0x00	0x00	0x00	0x00	0x00	0x00	0x00	0x00
+    ```
+    """
+    base_addr = gdb.parse_and_eval('$base("heap")')
+    return int(base_addr) + int(offset)
+
+
+@GdbFunction(only_when_running=True)
+def stack(offset: gdb.Value = gdb.Value(0)) -> int:
+    """
+    Return the base address of the stack mapping, optional offset allowed.
+
+    Example:
+    ```
+    pwndbg> vmmap
+    [...]
+    0x7ffff7fc5000     0x7ffff7fc6000 r--p     1000       0 /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
+    0x7ffff7fc6000     0x7ffff7ff1000 r-xp    2b000    1000 /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
+    0x7ffff7ff1000     0x7ffff7ffb000 r--p     a000   2c000 /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
+    0x7ffff7ffb000     0x7ffff7ffd000 r--p     2000   36000 /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
+    0x7ffff7ffd000     0x7ffff7fff000 rw-p     2000   38000 /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
+    0x7ffffffde000     0x7ffffffff000 rw-p    21000       0 [stack]
+    [...]
+    pwndbg> p/x $stack()
+    $1 = 0x7ffffffde000
+    pwndbg> x/gx $stack()
+    0x7ffffffde000:	0x0000000000000000
+    ```
+    """
+    base_addr = gdb.parse_and_eval('$base("stack")')
+    return int(base_addr) + int(offset)
+
+
+def _resolve_symbol_address(candidates: list[str]) -> int | None:
+    """
+    Try to resolve any of the given symbol names to an int address.
+    Returns the address, or None if none of them can be resolved.
+    """
+    for name in candidates:
+        for expr in (name, f"&{name}"):
+            try:
+                val = gdb.parse_and_eval(expr)
+            except gdb.error:
+                continue
+            return int(val)
+
+    return None
+
+
+@GdbFunction(only_when_running=True)
+def bss(offset: gdb.Value = gdb.Value(0)) -> int:
+    """
+    Return the address of the program's BSS start. Optional offset allowed.
+
+    Example:
+    ```
+    pwndbg> p/x $bss()
+    $1 = 0x555555558010
+    pwndbg> p/x $bss(0x10)
+    $2 = 0x555555558020
+    ```
+    """
+    bss_symbols = [
+        "__bss_start",
+        "_bss_start",
+        "__bss_start__",
+        "bss_start",
+        "_edata",
+    ]
+
+    base = _resolve_symbol_address(bss_symbols)
+    if base is None:
+        raise gdb.GdbError("$bss error: could not find a known bss symbol.")
+
+    return base + int(offset)
+
+
+@GdbFunction(only_when_running=True)
+def got(offset: gdb.Value = gdb.Value(0)) -> int:
+    """
+    Return the address of the Global Offset Table (GOT), optional offset allowed.
+
+    Example:
+    ```
+    pwndbg> p/x $got()
+    $1 = 0x7ffff7e02b80
+    pwndbg> p/x $got(0x8)
+    $2 = 0x7ffff7e02b88
+    ```
+    """
+    got_symbols = [
+        "_GLOBAL_OFFSET_TABLE_",
+        "__global_offset_table",
+        "__global_offset_table__",
+        "GOT",
+        "__got_start",
+    ]
+
+    base = _resolve_symbol_address(got_symbols)
```

Comment:
Pretty sure we have some logic to find got somewhere, we can probably reuse it here